README.md
![GitHub last commit](https://img.shields.io/github/last-commit/strongloop/oasgraph.svg?style=flat)
![Travis (.org)](https://img.shields.io/travis/strongloop/oasgraph.svg?style=flat)
![David](https://img.shields.io/david/strongloop/oasgraph.svg?style=flat)
[![npm](https://img.shields.io/npm/v/oasgraph.svg?style=flat)](https://www.npmjs.com/package/oasgraph)
[![Join the chat at https://gitter.im/oasgraph/Lobby](https://badges.gitter.im/oasgraph/Lobby.svg?style=flat)](https://gitter.im/oasgraph/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

# OASGraph

Turns APIs described by OpenAPI specifications (OAS) into GraphQL interfaces.

<img src="https://raw.githubusercontent.com/strongloop/oasgraph/master/docs/translation.png" alt="Overview of translation" width="600">

An example application exposing GitHub's public REST-like API via OASGraph can be found [here](https://oasgraph-github.mybluemix.net/) ([source code](https://github.ibm.com/apiharmony/oasgraph-oauth-github-example)).


## Getting started

* [Quickstart guide](./docs/tutorials/quickstart.md): Learn how to quickly spin up GraphQL wrappers using the OASGraph CLI.
* [Library tutorial](./docs/tutorials/watson.md): Learn how to use OASGraph as a library, and how to improve the resulting GraphQL wrappers using OAS `link` definitions.
* [LoopBack tutorial](./docs/tutorials/loopback_tutorial.md): Learn how to use OASGraph to create GraphQL wrappers for APIs created with LoopBack 4.


## Characteristics

* **Data-centric**
  The GraphQL interface is created around the data definitions in the given OAS, not around the endpoints, leading to a natural use of GraphQL.

  <img src="https://raw.githubusercontent.com/strongloop/oasgraph/master/docs/data-centric.png" alt="Example of data-centric design" width="600">

* **Nested data**
  [Links](https://github.com/OAI/OpenAPI-Specification/blob/OpenAPI.next/versions/3.0.md#linksObject) defined in the OAS are used to compose data definitions. Furthermore, hierarchical path structures can be used to nest data via the [`addSubOperations`](#options-addsuboperations) option.

  <img src="https://raw.githubusercontent.com/strongloop/oasgraph/master/docs/links.png" alt="Example of links resolution" width="600">

* **Automatic query resolution**
  Automatically generated resolvers translate (nested) GraphQL queries to API requests. Request results are translated back to GraphQL responses.

  <img src="https://raw.githubusercontent.com/strongloop/oasgraph/master/docs/resolution.png" alt="Example of query resolution" width="600">

* **Mutations**
  Non-safe, non-idempotent API operations (e.g., `POST`, `PUT`, `DELETE`) are translated to GraphQL [mutations](http://graphql.org/learn/queries/#mutations). Input payload is type-checked.

  <img src="https://raw.githubusercontent.com/strongloop/oasgraph/master/docs/mutations.png" alt="Example of mutation" width="600">

* **Authentication**
  OASGraph currently supports authentication via API Key and basic auth. OASGraph wraps secured endpoints into a `viewer`, which takes the API key / credentials as input.

  <img src="https://raw.githubusercontent.com/strongloop/oasgraph/master/docs/auth.png" alt="Example of authentication" width="600">

* **API Sanitation**
  Parts of an API that not compatible with GraphQL are automatically sanitized. For example, API parameters and data definition names with unsupported characters (e.g., `-`, `.`, `,`, `:`, `;`...) are removed. GraphQL queries are desanitized to correctly invoke the REST API and the responses are resanitized to create GraphQL-compliant results.

  <img src="https://raw.githubusercontent.com/strongloop/oasgraph/master/docs/sanitization.png" alt="Example of sanitation" width="300">

* **Custom request options** Provide headers and query parameters to send with every API request. This allows, for example, to handle authentication or tag requests from GraphQL.

* **Swagger and OpenAPI 3 support** OASGraph can handle both Swagger (OpenAPI specification 2.0) as well as OpenAPI specification 3.


## Usage
Install this package. Make sure to also have installed [`GraphQL.js`](https://github.com/graphql/graphql-js) (using `npm i graphql`), as it is a required peer-dependency. Then, simply pass it an OpenAPI Specification 3.0. The library returns a promise that resolves on an object containing the schema:

```javascript
const OASGraph = require('oasgraph') // use real name here

let oas = require('./fixtures/example_oas.json') // or other means of obtaining the OAS

OASGraph.createGraphQlSchema(oas)
  .then(({schema}) => {
    // do something with the schema
  })
  .catch(err => {
    // handle errors when creating the schema
  })
```

You can then use the generated schema, for example to be served using [Express.js](http://expressjs.com/):

```javascript
const express = require('express')
const graphqlHTTP = require('express-graphql')
const OASGraph = require('oasgraph') // use real name here
const app = express()

OASGraph.createGraphQlSchema(oas)
  .then(({schema}) => {
    app.use('/graphql', graphqlHTTP({
      schema,
      graphiql: true
    }))
    app.listen(3001)
  })
  .catch(err => {
    // handle errors when creating the schema
  })
```

***

Alternatively, use the CLI tool, which will run OASGraph on a specified OAS and start a server to host the GraphQL interface.

You can specify a local file containing the OAS specification or a remote url such as `http://127.0.0.1:3000/openapi.json`. Additionally, you can specify an optional port number so you can have multiple GraphQL servers for testing in the same machine.

```sh
oasgraph <OAS JSON file path or remote url> [port number]
```

To create the CLI tool, run:
```sh
npm link
```

Please note that the CLI tool is mainly used for quick testing and does not offer all the features that `createGraphQlSchema(oas, options)` does.

***

OASgraph can also generate and save the GraphQL schema to the local file `schema.graphql`, which you can use later to inspect or change its content. Please note that the following command will not start the GraphQL server.

```sh
oasgraph <OAS JSON file path or remote url> --save
```


## Options
OASGraph allows to define an optional `options` object:

```javascript
OASGraph.createGraphQLSchema(oas, options)
```

The following options can be set:

* `strict` (type: `boolean`, default: `false`): OASGraph generally tries to produce a working GraphQL interface for a given OAS. If OASGraph cannot fully translate a given OAS (e.g., because data schema definitions are incomplete or there are name collusions that cannot be resolved), OASGraph will per default degrade gracefully and produce a partly working GraphQL interface. OASGraph will log warnings (given logging is enabled). If OASGraph operates in `strict` mode, however, it will throw if it cannot create a GraphQL interface matching the given OAS perfectly.

* `headers` (type: `object`, default: `{}`): Headers to be sent in every request. Parameters defined in the OpenAPI Specification to set these headers will be ignored by OASGraph.

* `qs` (type: `object`, default: `{}`): Query parameters to be sent in every request. Parameters defined in the OpenAPI Specification to set these query parameters will be ignored by OASGraph.

* `viewer` (type: `boolean`, default: `true`): The viewer object types (i.e. QueryViewer and MutationViewer) are artificial constructs that allow a user to pass authentication credentials to OASGraph. Unfortunately, they are bulky and do not provide an accurate representation of the API. Depending on the API, it may be possible to send all your credentials through the header option, so if you would like to authenticate without the OASGraph-generated viewer object types, you can set the viewer option to false.

* `tokenJSONpath` (type: `string`, default: `undefined`): Used to pass the [JSONPath](http://goessner.net/articles/JsonPath/) of the OAuth token in the GraphQL context. To see more details, click [here](./README.md#authorization).

* <a name="options-addsuboperations"></a> `addSubOperations` (type: `boolean`, default: `false`): When true, OASGraph will nest `GET` operations based on their path hierarchy in the given OAS. E.g., when the OAS contains two paths `/users/{id}` and `/users/{id}/friends`, OASGraph will make `friends` queryable from within `user`. Note: This may cause problems when resolving GraphQL types in certain contexts, where the required variables are not available.

Consider this example of passing options:

```javascript
OASGraph.createGraphQLSchema(oas, {
  headers: {
    authorization: 'asfl3032lkj2' // send authorization header in every request
    'x-origin': 'GraphQL' // send header to identify requests made via GraphQL
  },
  qs: {
    limit: 30 // send limit query string in every request
  },
  addSubOperations: false
})
```


## Authentication
Per default, OASGraph will wrap API requests that need authentication in corresponding `viewers`, which allow the user to pass required credentials. OASGraph currently supports viewers for basic authentication and API keys. For example, a query using an API key viewer is:

```javascript
{
  viewerApiKey (apiKey: "api_key_here") {
    ...  // query for authenticated data here
  }
}
```

OASGraph uses dedicated viewers for mutations. For example, a mutation using a basic authentication viewer is:

```javascript
mutation {
  mutationViewerBasic (username: "user", password: "secret") {
    ...  // mutate authenticated data here
  }
}
```

OASGraph further provides `anyAuth` viewers (for queries and mutations), which allow the user to simultaneously provide information for multiple authentication mechanisms. AnyAuth viewers allow OASGraph to resolve nested queries and mutations that encompass API requests with different authentication mechanisms. For example, consider the following query:

```javascript
{
  viewerAnyAuth (
    exampleApiKeyProtocol: {apiKey: "a1p2i3k4e5y"}
    exampleBasicProtocol: {
      username: "erik"
      password: "secret"
    }
  ) {
    patentWithId (patentId: "test") {  // requires "exampleApiKeyProtocol"
      patentId
      inventor {                       // requires "exampleBasicProtocol"
        name
      }
    }
  }
}
```


## Authorization
OASGraph now supports OAuth 2.0!

Because OASGraph is a library, it cannot make the callbacks that OAuth requires by itself. Instead, the user must take care of the callback. After the user has obtained the OAuth token from the callback, simply pass the token, specifically the path of the token, to OASGraph through the `tokenJSONpath` [option](./README.md#options).

To see an example of how this would work, click [here](https://github.ibm.com/apiharmony/oasgraph-oauth-github-example)!


## Testing
To test OASGraph, run:

```bash
npm test
```

This command will temporarily start and later shut down an example REST(-like) API.


## Research
Our research paper, "Generating GraphQL-Wrappers for REST(-like) APIs", can be found [here](https://arxiv.org/abs/1809.08319). The paper describes the challenges of building OASGraph and an experiment in which we evaluated OASGraph against 959 publicly available OAS, provided by [APIs.guru](https://apis.guru/), and successfully created GraphQL interfaces for 89.5% of them.

To run the experiment, load APIs.guru specifications, found [here](https://github.com/APIs-guru/openapi-directory), into the `/tmp` folder:

```bash
npm run guru-load
```

Then, run tests:

```bash
npm run guru-test <number of APIs to test at most>
```


## Logging
OASGraph provides multiple levels of logging, which can be controlled by a `DEBUG` environment variable. You can enable these levels using:

```bash
DEBUG=level_1,level_2 node app-using-oasgraph.js
```

The following logging levels are supported:

* `preprocessing`: Logs information about preprocessing the OAS to GraphQL.
* `translation`: Logs information about translating an OAS to GraphQL.
* `http`: Logs information about the HTTP requests made to the API.


## Development
OASGraph is written in [TypeScript](http://www.typescriptlang.org/). All source code is contained in the `src` folder. Use `npm run build` or `npm test` to transpile the source files into the final library in the `lib` folder. Entry-point for the library is `index.js` in `lib`.


## Similar projects

* [swagger-to-graphql](https://github.com/yarax/swagger-to-graphql) turns a given Swagger (OpenAPI Specification 2.0) into a GraphQL interface, which resolves against the original API. GraphQL schema is based on endpoints, not on data definitions. No links are considered.

* [json-to-graphql](https://github.com/aweary/json-to-graphql) turns given JSON objects / arrays into a GraphQL schema. `resolve` functions need to be provided by the user.

* [StackOverflow discussion](https://stackoverflow.com/questions/38339442/json-schema-to-graphql-schema-converters) points to the above projects.


## License
[MIT](./LICENSE.md)


CHANGES.md
2018-10-10, Version 0.10.1
==========================

 * fixed cli with file path (Erik Wittern)


2018-10-08, Version 0.10.0
==========================

 * Peer dependency mistake (Alan Cha)

 * Fixed some problems with the documentation (Alan Cha)

 * feat(doc) update new features (Mario Estrada)

 * Set up Travis CI (Alan Cha)

 * Added tutorial video (Alan Cha)

 * feat(cli) add option to save schema (Mario Estrada)

 * Programmatically run example API during tests to address #39 (Erik Wittern)

 * refactor(cli) add support for remote spec file (Mario Estrada)

 * Added new tutorials (Alan Cha)

 * Reverted CLI integration (Alan Cha)

 * Revert "add travis" (Alan Cha)

 * Revert "use node 8 and above for jenkins" (Alan Cha)

 * use node 8 and above for jenkins (Diana Lau)

 * add travis (Diana Lau)

 * Removed linking from and change badge in the README (Alan Cha)

 * Integrated CLI tool into the main entry point file (Alan Cha)

 * Modified .gitignore and updated dependencies, fixed #28 (Alan Cha)

 * Added reference to research paper and badges (Alan Cha)

 * Changed back to absolute links, added image for sanitation, reworded parts of the quick start guide (Alan Cha)


2018-09-19, Version 0.9.1
=========================



2018-09-14, Version 0.9.0
=========================

 * Refactored functions (Alan Cha)

 * Added support for header parameters and cookies, related #20, fixed some naming issues associated with 9fc0273 (Alan Cha)

 * Small title change (Alan Cha)

 * Renamed file for consistency (Alan Cha)

 * Added quick start guide (Alan Cha)

 * Initial version (laredoj)

 * Added very basic cli tool (Alan Cha)


2018-09-04, Version 0.8.0
=========================

 * Removed real names from test data, fixed #87 (Alan Cha)

 * Added support for non-application/json request and support bodies (Alan Cha)

 * Delete logo.png (JOHN ERIK E. WITTERN)

 * Update README.md (JOHN ERIK E. WITTERN)

 * Delete oasgraph.pptx (JOHN ERIK E. WITTERN)

 * Deleted spurious line #86 (Alan Cha)

 * Update package.json (JOHN ERIK E. WITTERN)


2018-08-24, Version 0.7.1
=========================

 * Fixed relative links (Alan Cha)

 * Change absolute links to relative links (Alan Cha)

 * Added licensing (Alan Cha)


2018-08-23, Version 0.7.0
=========================

 * Supports 2XX status code (Alan Cha)

 * Updated dependencies, modified example OAS to comply with faulty dependency, now it does not comply with OAS 3.0.0, will update as soon as the dependency fixes issue (Alan Cha)

 * Added operationRef functionality and tests, fixed old tests, close #5 (Alan Cha)


2018-08-15, Version 0.6.1
=========================

 * Added missing devdependency (Alan Cha)


2018-08-10, Version 0.6.0
=========================

 * reflect changed beyond simple patch (Erik Wittern)

 * updated version to reflect changed dependencies (Erik Wittern)

 * replaced nodegit with isomorphic-git to fix #81 (Erik Wittern)

 * setup linting via Standard.js and linted all source files (Erik Wittern)

 * extended example API with GET /users endpoint (Erik Wittern)

 * removed unneeded ignore statements (Erik Wittern)

 * further refactorings towards #80 (Erik Wittern)

 * refactoring of index (Erik Wittern)

 * revised scripts to work with TypeScript (Erik Wittern)

 * basic replacement of Flow with TypeScript towards #80 (Erik Wittern)


2018-03-21, Version 0.5.2
=========================

 * added tests to ultimately fix #78 (Erik Wittern)

 * centering (Erik Wittern)

 * better logo placement (Erik Wittern)

 * added logo (Erik Wittern)


2018-01-30, Version 0.5.1
=========================

 * do not lowercase first char of enum values; bumped version to 0.5.1 (Erik Wittern)

 * updated evaluation results (Erik Wittern)


2018-01-25, Version 0.5.0
=========================

 * use external library to resolve allOf definitions to fix #77; bump minor version accordingly (Erik Wittern)


2018-01-25, Version 0.4.0
=========================

 * added support for int64 formatted integers to fix #76; updated minor version correspondingly (Erik Wittern)


2018-01-24, Version 0.3.1
=========================

 * fixed error where viewer option set to false was not considered properly; bumped patch version correspondingly (Erik Wittern)


2018-01-24, Version 0.3.0
=========================

 * added support for arbitrary JSON types to fix #75; bumped version to 0.3.0 correspondingly (Erik Wittern)

 * made GraphQL.js peer dependency to fix #74 (Erik Wittern)

 * added evaluation results (Erik Wittern)


2018-01-16, Version 0.2.1
=========================



2018-01-16, Version 0.1.1
=========================

 * fixed addSubOperations option; improved logging for Object Type creation; improved APIs.guru eval (Erik Wittern)

 * create suboperations only when option is set to true (Erik Wittern)

 * added further analysis of eval results (Erik Wittern)

 * fixed example GraphQL server (Erik Wittern)

 * revised evaluation script for better performance (Erik Wittern)

 * added reporting of number of successfully created queries / mutations (Erik Wittern)

 * extended comments documenting creation of placeholders (Erik Wittern)


2018-01-09, Version 0.2.0
=========================

 * minor version bump (Erik Wittern)

 * customize the names used for placeholders to fix #72 (Erik Wittern)

 * mark "query" and "mutation" as already used to fix #73 (Erik Wittern)

 * fixed issue #71 where names would collide based on convention of adding Input (Erik Wittern)

 * refactored evaluation code (Erik Wittern)

 * provided classification of errors to evaluation (Erik Wittern)

 * major revision of warnings format; basis for evaluation (Erik Wittern)

 * revised evaluation scripts (Erik Wittern)

 * updated flow versions; minor related fixes (Erik Wittern)

 * added built utils (Erik Wittern)

 * refactoring of warnings (Erik Wittern)


2017-11-28, Version 0.1.1
=========================


lib/index.js
"use strict";
// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: oasgraph
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
// Imports:
const schema_builder_1 = require("./schema_builder");
const resolver_builder_1 = require("./resolver_builder");
const GraphQLTools = require("./graphql_tools");
const preprocessor_1 = require("./preprocessor");
const Oas3Tools = require("./oas_3_tools");
const auth_builder_1 = require("./auth_builder");
const debug_1 = require("debug");
const log = debug_1.default('translation');
/**
 * Creates a GraphQL interface from the given OpenAPI Specification (2 or 3).
 */
function createGraphQlSchema(spec, options) {
    return __awaiter(this, void 0, void 0, function* () {
        // deal with option defaults:
        // @ts-ignore
        if (typeof options === 'undefined')
            options = {};
        options.strict = typeof options.strict === 'boolean'
            ? options.strict
            : false;
        options.addSubOperations = typeof options.addSubOperations === 'boolean'
            ? options.addSubOperations
            : false;
        options.viewer = typeof options.viewer === 'boolean'
            ? options.viewer
            : true;
        options.sendOAuthTokenInQuery = typeof options.sendOAuthTokenInQuery === 'boolean'
            ? options.sendOAuthTokenInQuery
            : false;
        options.report = {
            warnings: [],
            numOps: 0,
            numOpsQuery: 0,
            numOpsMutation: 0,
            numQueriesCreated: 0,
            numMutationsCreated: 0
        };
        /**
         * Check if the spec is a valid OAS 3.0.x
         * If the spec is OAS 2.0, attempt to translate it into 3.0.x, then try to
         * translate the spec into a GraphQL schema
         */
        let oas = yield Oas3Tools.getValidOAS3(spec);
        let { schema, report } = yield translateOpenApiToGraphQL(oas, options);
        return {
            schema,
            report
        };
    });
}
exports.createGraphQlSchema = createGraphQlSchema;
/**
 * If this module is run as a command line tool
 *
 * Usage: oasgraph <OAS JSON file path>
 *
 * Creates a server hosting the GraphQL interface generated form the given
 * OpenAPI Specification (2 or 3).
 *
 * NOTE: Mainly used for quick testing. Does not provide the flexibility and
 * functionality of createGraphQlSchema(oas, options)
 */
if (require.main === module) {
    const express = require('express');
    const graphqlHTTP = require('express-graphql');
    const path = require('path');
    const app = express();
    const [, , ...args] = process.argv;
    // Check for arguments
    if (args.length > 0) {
        let filePath = args[0];
        // Convert path to absolute path
        // Otherwise, would need to have separate cases for relative or absolute paths
        filePath = path.resolve(filePath);
        try {
            let oas = require(filePath);
            // Create GraphQL interface
            createGraphQlSchema(oas, { strict: false, viewer: true, addSubOperations: true, sendOAuthTokenInQuery: false, report: {
                    warnings: [],
                    numOps: 0,
                    numOpsQuery: 0,
                    numOpsMutation: 0,
                    numQueriesCreated: 0,
                    numMutationsCreated: 0
                } }).then(({ schema, report }) => {
                console.log(JSON.stringify(report, null, 2));
                app.use('/graphql', graphqlHTTP({
                    schema: schema,
                    graphiql: true
                }));
                app.listen(3001, () => {
                    console.log('GraphQL accessible at: http://localhost:3001/graphql');
                });
            })
                .catch(err => {
                console.log(err);
            });
        }
        catch (error) {
            if (error.code === 'MODULE_NOT_FOUND') {
                console.log(`The file path "${filePath}" is invalid.`);
            }
            else {
                console.log(error);
            }
        }
    }
    else {
        console.log('Usage: oasgraph <OAS JSON file path>');
    }
}
/**
 * Creates a GraphQL interface from the given OpenAPI Specification 3.0.x
 */
function translateOpenApiToGraphQL(oas, { strict, headers, qs, viewer, tokenJSONpath, addSubOperations, sendOAuthTokenInQuery, report }) {
    return __awaiter(this, void 0, void 0, function* () {
        let options = {
            headers,
            qs,
            viewer,
            tokenJSONpath,
            strict,
            addSubOperations,
            sendOAuthTokenInQuery,
            report
        };
        log(`Options: ${JSON.stringify(options)}`);
        /**
         * Extract information from the OAS and put it inside a data structure that
         * is easier for OASGraph to use
         */
        let data = preprocessor_1.preprocessOas(oas, options);
        /**
         * Create GraphQL fields for every operation and structure them based on their
         * characteristics (query vs. mutation, auth vs. non-auth).
         */
        let queryFields = {};
        let mutationFields = {};
        let authQueryFields = {};
        let authMutationFields = {};
        Object.entries(data.operations)
            // Start with endpoints that DO contain links OR that DO contain sub
            // operations, so that built-up GraphQL object types contain these links
            // when they are re-used.
            .sort(([op1Id, op1], [op2Id, op2]) => sortByHasLinksOrSubOps(op1, op2))
            .forEach(([operationId, operation]) => {
            log(`Process operation "${operationId}"...`);
            let field = getFieldForOperation(operation, data, oas);
            if (!operation.isMutation) {
                let fieldName = operation.responseDefinition.otName;
                if (operation.inViewer) {
                    for (let securityRequirement of operation.securityRequirements) {
                        if (typeof authQueryFields[securityRequirement] !== 'object') {
                            authQueryFields[securityRequirement] = {};
                        }
                        // Avoid overwriting fields that return the same data:
                        if (fieldName in authQueryFields[securityRequirement]) {
                            fieldName = Oas3Tools.beautifyAndStore(operationId, data.saneMap);
                        }
                        authQueryFields[securityRequirement][fieldName] = field;
                    }
                }
                else {
                    // Avoid overwriting fields that return the same data:
                    if (fieldName in queryFields) {
                        fieldName = Oas3Tools.beautifyAndStore(operationId, data.saneMap);
                    }
                    queryFields[fieldName] = field;
                }
            }
            else {
                // Use operationId to avoid problems differentiating operations with the
                // same path but differnet methods
                let saneFieldName = Oas3Tools.beautifyAndStore(operationId, data.saneMap);
                if (operation.inViewer) {
                    for (let securityRequirement of operation.securityRequirements) {
                        if (typeof authMutationFields[securityRequirement] !== 'object') {
                            authMutationFields[securityRequirement] = {};
                        }
                        authMutationFields[securityRequirement][saneFieldName] = field;
                    }
                }
                else {
                    mutationFields[saneFieldName] = field;
                }
            }
        });
        /**
         * Count created queries / mutations
         */
        options.report.numQueriesCreated =
            Object.keys(queryFields).length +
                Object.keys(authQueryFields).reduce((sum, key) => {
                    return sum + Object.keys(authQueryFields[key]).length;
                }, 0);
        options.report.numMutationsCreated =
            Object.keys(mutationFields).length +
                Object.keys(authMutationFields).reduce((sum, key) => {
                    return sum + Object.keys(authMutationFields[key]).length;
                }, 0);
        /**
         * Organize created queries / mutations into viewer objects.
         */
        if (Object.keys(authQueryFields).length > 0) {
            Object.assign(queryFields, auth_builder_1.createAndLoadViewer(authQueryFields, data, oas, false));
        }
        if (Object.keys(authMutationFields).length > 0) {
            Object.assign(mutationFields, auth_builder_1.createAndLoadViewer(authMutationFields, data, oas, true));
        }
        /**
         * Build up the schema
         */
        const schemaConfig = {
            query: Object.keys(queryFields).length > 0
                ? new graphql_1.GraphQLObjectType({
                    name: 'query',
                    description: 'The start of any query',
                    fields: queryFields
                })
                : GraphQLTools.getEmptyObjectType('query'),
            mutation: Object.keys(mutationFields).length > 0
                ? new graphql_1.GraphQLObjectType({
                    name: 'mutation',
                    description: 'The start of any mutation',
                    fields: mutationFields
                })
                : null
        };
        // Fill in yet undefined Object Types to avoid GraphQLSchema from breaking.
        // The reason: once creating the schema, the 'fields' thunks will resolve
        // and if a field references an undefined Object Types, GraphQL will throw.
        Object.entries(data.operations).forEach(([opId, operation]) => {
            if (typeof operation.responseDefinition.ot === 'undefined') {
                operation.responseDefinition.ot = GraphQLTools
                    .getEmptyObjectType(operation.responseDefinition.otName);
            }
        });
        const schema = new graphql_1.GraphQLSchema(schemaConfig);
        return { schema, report: options.report };
    });
}
/**
 * Helper function for sorting operations based on them having links or sub-
 * operations.
 */
function sortByHasLinksOrSubOps(op1, op2) {
    const hasOp1 = Object.keys(op1.links).length > 0 ||
        (Array.isArray(op1.subOps) && op1.subOps.length > 0);
    const hasOp2 = Object.keys(op2.links).length > 0 ||
        (Array.isArray(op2.subOps) && op2.subOps.length > 0);
    return (hasOp1 === hasOp2) ? 0 : hasOp1 ? -1 : 1; // hasOp1 = true => -1 = first
}
/**
 * Creates the field object for the given operation.
 */
function getFieldForOperation(operation, data, oas) {
    // create GraphQL Type for response:
    let type = schema_builder_1.getGraphQLType({
        name: operation.responseDefinition.otName,
        schema: operation.responseDefinition.schema,
        data,
        operation,
        oas
    });
    // create resolve function:
    let payloadSchemaName = operation.payloadDefinition
        ? operation.payloadDefinition.iotName
        : null;
    let payloadSchema = operation.payloadDefinition
        ? operation.payloadDefinition.schema
        : null;
    let resolve = resolver_builder_1.getResolver({
        operation,
        oas,
        payloadName: payloadSchemaName,
        data
    });
    // create args:
    let args = schema_builder_1.getArgs({
        parameters: operation.parameters,
        payloadSchemaName: payloadSchemaName,
        payloadSchema,
        operation,
        data,
        oas
    });
    return {
        type,
        resolve,
        args,
        description: operation.description
    };
}
//# sourceMappingURL=index.js.map